### PR TITLE
Ensure reader leaves consumer group on graceful exit

### PR DESCRIPTION
**Pull request by stevevls**
Issue #164

### CODE REVIEW

File: reader.go [hunk 1]
```diff
@@ -1161,7 +1161,18 @@ func (r *Reader) Close() error {
...
+			// note that we do not call r.coordinator() because that passes
+			// r.stctx go DialContext.  we just cancelled that context a couple
+			// lines above by calling r.stop().
+			r.mutex.Lock()
+			address := r.address
+			r.mutex.Unlock()
+
+			// DialContext will wrap the background context to add appropriate
+			// timeout/deadline configuration.
+			conn, err := r.config.Dialer.DialContext(context.Background(), "tcp", address)
```
Comment (stevevls) [inline reader.go]: @achille-roussel can you 2x check that this is a correct assumption?

Comment (achille-roussel) [reply reader.go]: That's assuming there is a timeout configured on the dialer, but if the program sets none it's definitely looking for trouble so I think we're good there!

**Comment by WillAbides** [comment]
I was working on a PR as well.  For the fix, I just moved the whole `if r.useConsumerGroup()` block to the top of the `Close()` method.  I don't know if that method or this one will work better.

Can I suggest adding the test below?  It passes for the changes in this PR.

```go
func TestCloseLeavesGroup(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	topic := makeTopic()
	createTopic(t, topic, 1)
	r := NewReader(ReaderConfig{
		Brokers:  []string{"localhost:9092"},
		Topic:    topic,
		GroupID:  makeGroupID(),
		MinBytes: 1,
		MaxBytes: 10e6,
		MaxWait:  100 * time.Millisecond,
	})
	prepareReader(t, ctx, r)
	groupID := r.Config().GroupID

	// wait for generationID > 0 so we know our reader has joined the group
	membershipTimer := time.After(1 * time.Second)
	for {
		done := false
		select {
		case <-membershipTimer:
			t.Fatalf("our reader never joind its group")
		default:
			generationID, _ := r.membership()
			if generationID > 0 {
				done = true
			}
		}
		if done {
			break
		}
	}

	err := r.Close()
	if err != nil {
		t.Fatalf("unexpected error closing reader: %s", err.Error())
	}

	conn, err := Dial("tcp", "localhost:9092")
	if err != nil {
		t.Fatalf("error dialing: %v", err)
	}
	resp, err := conn.describeGroups(describeGroupsRequestV1{
		GroupIDs: []string{groupID},
	})
	if err != nil {
		t.Fatalf("error from describeGroups %v", err)
	}
	if len(resp.Groups) != 1 {
		t.Fatalf("expected 1 group. got: %d", len(resp.Groups))
	}
	if len(resp.Groups[0].Members) != 0 {
		t.Fatalf("expected group membership size of %d, but got %d", 0, len(resp.Groups[0].Members))
	}
}
```

**Comment by stevevls** [comment]
@achille-roussel I agree re: the connection.  Let me dig a little further.  We may be able to add a closing check in `handshake` and leave the group then.  We def. don't want to just call `leaveGroup` because `handshake` is called once per generation...that could result in all consumers leaving the group when one does.

@WillAbides I don't think that moving the code up is a good idea because we should wait for the reader go routines to exit before leaving the group.  The unit test looks great, though.  Feel free to commit it to this branch!

**Comment by stevevls** [comment]
Just pushed code that moves leaving the group into the `handshake` function.  It checks to see if the reader's context is done and leaves the group if so.  The reader's context isn't used by the call graph when invoking `leaveGroup` function, so we don't run into the original issue.  The update passes the unit test that @WillAbides provided.  I didn't check in the test, though, because I prefer that he gets recognition for contributing it.

**Comment by WillAbides** [comment]
I don't have rights to add to this branch directly, but I created #190 to add the test to this branch.

**Comment by WillAbides** [comment]
That test passed before 2cead73, but it's failing now.

**Comment by stevevls** [comment]
@WillAbides I think there's a timing condition in the test.  Adding a short sleep after the `membershipTimer` receive and before the `r.Close()` makes the test pass.  I think the issue is that we don't have a connection with which we can leave the group until after calling `rebalance` on line 794.  However, I'm seeing that the call to `rebalance` is error-ing out if the reader is closed before it finishes.  I may be able to refactor this so that we get a handle to the connection earlier.

**Comment by stevevls** [comment]
@WillAbides Give this latest commit a try.  I refactored it so that all of the consumer group operations share a single `Conn`.  By establishing the connection first, we can then put the `defer` block to leave the group if any of the CG operations fail.  The test you provided was passing reliably for me against a local kafka 0.11.0 cluster.

**Comment by WillAbides** [comment]
@stevevls It's also passing for me using the kafka provided by docker-compose.

**Comment by stevevls** [comment]
Cool.  So if this looks okay to @achille-roussel, we can merge it in.

**Pull request merged**